File: src/main/scala/actor/PreloaderActor.scala
package actor

import akka.actor.{ActorRef, Actor}
import akka.pattern.ask
import akka.util.Timeout

import com.mongodb.casbah.Imports._
import scala.concurrent.Future

import base._
import scala.slick.driver.SQLiteDriver.simple._
import file.Location


class PreloaderActor extends Actor {

  implicit val timeout = Timeout(120000)

  import context.dispatcher

  private val artistLoader = context.actorFor("akka://LoadingSystem/user/artistLoader")
  private val fileReader = context.actorFor("akka://LoadingSystem/user/fileReader")

  private val songReader = context.actorFor("akka://LoadingSystem/user/songReader")
  private val termOrTagReader = context.actorFor("akka://LoadingSystem/user/termOrTagReader")
  private val similaritiesReader = context.actorFor("akka://LoadingSystem/user/similaritiesReader")

  private val locationsCollection = context.actorFor("akka://LoadingSystem/user/locations")
  private val similarititesCollection = context.actorFor("akka://LoadingSystem/user/similaritites")
  private val songsCollection = context.actorFor("akka://LoadingSystem/user/songs")
  private val tagsCollection = context.actorFor("akka://LoadingSystem/user/tags")
  private val termsCollection = context.actorFor("akka://LoadingSystem/user/terms")

  private val progressListener = context.actorFor("akka://LoadingSystem/user/progressListener")

  def receive = {
    case Go => {

      progressListener ! StartListener(5)

      val locations: Future[Message] = loadLocations()

      for {
        _ <- loadSongs()
        _ <- loadSimilarities()
        _ <- loadTags()
        _ <- loadTerms()
        _ <- locations
      }
      yield (artistLoader.tell(Go, sender = self))
    }
  }

  def loadLocations(): Future[Message] = {
    val locations = ask(fileReader, LoadFile("subset_artist_location.txt"))
      .mapTo[FileLoaded]
      .map(fileLoaded => fileLoaded.lines.map(Location.apply(_)))

    ProgressListener("locations", progressListener)(for {
      beans <- locations
      status <- ask(locationsCollection, Write(beans)).mapTo[Message]
    } yield (status))
  }

  private def loadSongs(): Future[Message] = {
    load("songs", songReader, songsCollection) {
      implicit session => {
        val query = for (song <- Songs) yield (song)
        query.mapResult(_.toMongo).list()
      }
    }
  }

  private def loadSimilarities(): Future[Message] = {
    load("similarities", similaritiesReader, similarititesCollection) {
      implicit session => {
        val query = for (artist <- ArtistSimilarities) yield (artist)
        query.mapResult(_.toMongo).list()
      }
    }
  }

  private def loadTags(): Future[Message] = {
    load("tags", termOrTagReader, tagsCollection) {
      implicit session => {
        val query = for (tagByArtist <- TagsByArtist) yield (tagByArtist)
        query.mapResult(_.toMongo).list()
      }
    }
  }

  private def loadTerms(): Future[Message] = {
    load("terms", termOrTagReader, termsCollection) {
      implicit session => {
        val query = for (temByArtist <- TermsByArtist) yield (temByArtist)
        query.mapResult(_.toMongo).list()
      }
    }
  }

  private def load(message: String, reader: ActorRef, writer: ActorRef)(all: (Session) => List[MongoDBObject]): Future[Message] = {
    ProgressListener(message, progressListener)(for {
      beans <- ask(reader, Extract(all)).mapTo[Extracted[List[MongoDBObject]]]
      status <- ask(writer, Write(beans.data)).mapTo[Message]
    }
    yield (status))
  }
}

File: src/main/java/actor/PreloaderActor.java
package actor;

import actor.message.*;
import akka.actor.ActorRef;
import akka.actor.UntypedActor;
import akka.dispatch.Future;
import akka.dispatch.Mapper;
import akka.dispatch.OnComplete;
import akka.dispatch.OnSuccess;
import akka.util.Timeout;
import base.ArtistSimilarity;
import base.Song;
import base.TagByArtist;
import base.TermByArtist;
import file.Location;

import java.util.List;

import static base.ArtistSimilarity.toArtistSimilarity;
import static base.Song.toSong;
import static base.TagByArtist.toTagByArtist;
import static com.google.common.collect.FluentIterable.from;
import static com.google.common.collect.Lists.newArrayList;
import static file.Location.toLocation;
import static akka.pattern.Patterns.ask;
import static akka.dispatch.Futures.sequence;


public class PreloaderActor extends UntypedActor {

    private static final Timeout timeout = Timeout.apply(120000);

    private ActorRef artistLoader = getContext().actorFor("akka://LoadingSystem/user/artistLoader");
    private ActorRef fileReader = getContext().actorFor("akka://LoadingSystem/user/fileReader");

    private ActorRef songReader = getContext().actorFor("akka://LoadingSystem/user/songReader");
    private ActorRef termOrTagReader = getContext().actorFor("akka://LoadingSystem/user/termOrTagReader");
    private ActorRef similaritiesReader = getContext().actorFor("akka://LoadingSystem/user/similaritiesReader");

    private ActorRef locationsCollection = getContext().actorFor("akka://LoadingSystem/user/locations");
    private ActorRef similaritiesCollection = getContext().actorFor("akka://LoadingSystem/user/similaritites");
    private ActorRef songsCollection = getContext().actorFor("akka://LoadingSystem/user/songs");
    private ActorRef tagsCollection = getContext().actorFor("akka://LoadingSystem/user/tags");
    private ActorRef termsCollection = getContext().actorFor("akka://LoadingSystem/user/terms");

    private ActorRef progressListener = getContext().actorFor("akka://LoadingSystem/user/progressListener");

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof Go) {
            progressListener.tell(new StartListener(5));

            Future<Message> locations = loadLocations();
            Future<Message> details = loadSongs()
                    .flatMap(new Mapper<Message, Future<Message>>() {
                        @Override
                        public Future<Message> apply(Message parameter) {
                            return loadSimilarities();
                        }
                    }).flatMap(new Mapper<Message, Future<Message>>() {
                        @Override
                        public Future<Message> apply(Message parameter) {
                            return loadTerms();
                        }
                    }).flatMap(new Mapper<Message, Future<Message>>() {
                        @Override
                        public Future<Message> apply(Message parameter) {
                            return loadTags();
                        }
                    });

            Future<Iterable<Message>> status = sequence(newArrayList(locations, details), getContext().dispatcher());

            status.onSuccess(new OnSuccess<Iterable<Message>>() {
                @Override
                public void onSuccess(Iterable<Message> result) throws Throwable {
                    artistLoader.tell(Go.singleton);
                }
            });

        } else {
            unhandled(message);
        }
    }

    private Future<Message> loadLocations() {
        Future<Write> locations = ask(fileReader, new LoadFile("subset_artist_location.txt"), timeout)
                .map(new Mapper<Object, Write>() {
                    @Override
                    public Write apply(Object input) {
                        List<String[]> sources = (List<String[]>) input;

                        return new Write(from(sources)
                                .transform(toLocation())
                                .toArray(Location.class)
                        );
                    }
                });

        return locations.flatMap(new Mapper<Write, Future<Message>>() {
            @Override
            public Future<Message> apply(Write input) {
                return ask(locationsCollection, input, timeout)
                        .map(toMessage())
                        .onComplete(toProgressListener("locations"));
            }
        });
    }

    private Future<Message> loadSongs() {
        return load(songReader, new Extract<Song>("select * from songs", toSong()), songsCollection, "songs");
    }

    private Future<Message> loadSimilarities() {
        return load(similaritiesReader, new Extract<ArtistSimilarity>("select * from similarity", toArtistSimilarity()), similaritiesCollection, "similarities");
    }

    private Future<Message> loadTags() {
        return load(termOrTagReader, new Extract<TagByArtist>("select * from artist_mbtag", toTagByArtist()), tagsCollection, "tags");
    }

    private Future<Message> loadTerms() {
        return load(termOrTagReader, new Extract<TermByArtist>("select * from artist_term", TermByArtist.toTermByArtist()), termsCollection, "terms");
    }


    private Future<Message> load(ActorRef reader, Extract extract, final ActorRef writer, final String message) {
        Future<Write> beans = ask(reader, extract, timeout)
                .map(new Mapper<Object, Write>() {
                    @Override
                    public Write apply(Object input) {
                        return new Write(((Extracted) input).data.toArray());
                    }
                });

        return beans.flatMap(new Mapper<Write, Future<Message>>() {
            @Override
            public Future<Message> apply(Write input) {
                return ask(writer, input, timeout)
                        .map(toMessage())
                        .onComplete(toProgressListener(message));
            }
        });
    }

    private OnComplete<Message> toProgressListener(final String successfulMessage) {
        return new OnComplete<Message>() {
            @Override
            public void onComplete(Throwable failure, Message success) throws Throwable {
                if (failure != null) {
                    progressListener.tell(failure);
                } else {
                    progressListener.tell(new Done(successfulMessage));
                }
            }
        };
    }

    public static Mapper<Object, Message> toMessage() {
        return new Mapper<Object, Message>() {
            @Override
            public Message apply(Object input) {
                return (Message) input;
            }
        };
    }

}
